app.ts

```diff
@@ -0,0 +1,61 @@
+/**  Debounce utility function to limit function execution frequency*/
+function debounce<F extends (...args: any) => any>(func: F, waitFor: number) {
+	let timeout: number;
+
+	return (...args: Parameters<F>): Promise<ReturnType<F>> => {
+		clearTimeout(timeout);
+
+		return new Promise((resolve) => {
+			timeout = setTimeout(() => {
+				resolve(func(...args));
+			}, waitFor);
+		});
+	};
+}
+
+/**  Constants for grid calculation
+ * GRID_RATIO represents the ratio of the grid's height to the window's height.
+*/
+const GRID_RATIO = 9/20 ;
+const ROW_HEIGHT = 16;
+
+/**  Wait for the document to be ready*/
+$(document).ready(() => {
+	// Initialization and setup code
+	const initialGridSize = Math.floor((<number>($(window).innerHeight()) * GRID_RATIO) / ROW_HEIGHT) ;
+	const apidata = new ApiData(initialGridSize);
+	// Set up search button click handler
+	$('#searchBtn').on('click', () => {
+
+		const from = parseInt(<string>($('#fromInput').val()));
+		const pageSize = apidata.pageSize;
+		const maxRange = apidata.totalItems - 1;
+		if (!isNaN(from) && from >= 0 && from <= maxRange) {
+			let to = Math.min(from + pageSize, maxRange);
+			let adjustedFrom = from;
+			if (adjustedFrom + pageSize > maxRange) {
+				adjustedFrom = Math.max(0, maxRange - pageSize);
+				to = maxRange;
+			}
+			apidata.searchRecords(adjustedFrom);
+		} else if (from < 0 || from > maxRange) {
+			alert('please enter values in the range (0-999999)');
```

Comment:
Not all back-ends will have the same max id. You will have to change this to not be hard coded.